### PR TITLE
Stack and concatenate routines including subroutines

### PR DESCRIPTION
**Pull request by sbak5**
Implemented concatenate and stack w/ subroutines which uses concatenate. 
Added tests for stack, concatenate and stack subroutines (vstack, hstack, and stack)
np.astype is changed a bit to handle casting in concatenate

### CODE REVIEW

File: tests/stack.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def run_test(arr, test_routine):
+    input_arr = []
+    for axis in range(arr[0].ndim):
+        input_arr.append(axis)
+
+    for routine, input_opt in itertools.product(test_routine, input_arr):
+        print_msg = f"np.{routine}((array({arr[0].shape}) * 3), {input_opt}"
+        b = getattr(np, routine)(arr, input_opt)
```
Comment (magnatelee) [inline tests/stack.py:31]: I feel this is an unnecessary complication, knowing that you only test `np.concatenate` in this code; this just makes it difficult to track what's being tested here.

File: tests/stack.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+def test(min_size, max_size):
+    if min_size >= max_size:
+        max_size = min_size + 1
+    print("test np.stack")
+    test_routine = ["stack"]
+    # test np.stack w/ 1D, 2D and 3D arrays
+    # test 1D arrays
+    for i in range(min_size, max_size):
+        a = tuple([np.random.randn(i) for num_arr in range(3)])
```
Comment (magnatelee) [inline tests/stack.py:68]: You can use a for comprehension directly in a tuple and brackets are unnecessary here.

File: tests/stack.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        run_test(a, test_routine)
+
+        # test 2D arrays
+        for j in range(min_size, max_size):
+            b = tuple([np.random.randn(i, j) for num_arr in range(3)])
```
Comment (magnatelee) [inline tests/stack.py:73]: same here (I won't add any more comments like this)

File: tests/stack_functions.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+import numpy as np
+
+import cunumeric as num
+
+
+def run_test(arr, test_routine):
+    for routine in test_routine:
+        print_msg = f"np.{routine}((array({arr[0].shape})*3)"
+        b = getattr(np, routine)(arr)
```
Comment (magnatelee) [inline tests/stack_functions.py:26]: Looks like you copy-pasted the test for `concatenate` from this one... is there a reason that you didn't test `concatenate` together in this test? you may want to consider merging the tests if possible.

File: cunumeric/array.py [hunk 5]
```diff
@@ -945,8 +945,23 @@ def astype(
...
+
+        casting_allowed = np.can_cast(self.dtype, dtype, casting)
+        if casting_allowed:
+            # For numeric to non-numeric casting, the dest dtype should be
+            # retrived from 'promote_types' to preserve values
+            # e.g. ) float 64 to str, np.dtype(dtype) == '<U'
+            # this dtype is not safe to store
+            if dtype == np.dtype("str"):
+                dtype = np.promote_types(self.dtype, dtype)
+            result = ndarray(self.shape, dtype=dtype, inputs=(self,))
```
[the remark this answers is not in the document]
Comment (sbak5) [reply cunumeric/array.py:957]: All right. You mean using `dtype` for both conditional statement and assignment after that is risky and error-prone?

File: cunumeric/module.py [hunk 6]
```diff
@@ -644,6 +644,195 @@ def array_split(a, indices, axis=0, equal=False):
...
+
+
+@copy_docstring(np.concatenate)
+def concatenate(
+    inputs,
+    axis=0,
+    out=None,
+    dtype=None,
+    casting="same_kind",
+    method_name=None,
```
[the remark this answers is not in the document]
Comment (sbak5) [reply cunumeric/module.py:797]: OK. I did this with reference to other top-level methods in `module.py`. I would do this as you suggsted.

File: cunumeric/array.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (magnatelee) [reply cunumeric/array.py:957]: no I'm talking about `result` being conditionally assigned

**Pull request merged**